File: app/src/main/java/miui/statusbar/lyric/view/LyricTextSwitchView.java
package miui.statusbar.lyric.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Typeface;
import android.text.TextPaint;
import android.text.TextUtils;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.ViewFlipper;

@SuppressLint("ViewConstructor")
public class LyricTextSwitchView extends ViewFlipper {

    LyricTextView lyricTextView;
    LyricTextView lyricTextView2;
    AutoMarqueeTextView autoMarqueeTextView;
    AutoMarqueeTextView autoMarqueeTextView2;
    boolean switchLyric = false;
    boolean hasMeizu;

    public LyricTextSwitchView(Context context, boolean isMeizu) {
        super(context);
        hasMeizu = isMeizu;
        if (isMeizu) {
            lyricTextView = new LyricTextView(context);
            lyricTextView2 = new LyricTextView(context);

            lyricTextView.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT));
            lyricTextView2.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT));

            this.addView(lyricTextView);
            this.addView(lyricTextView2);
        } else {
            autoMarqueeTextView = new AutoMarqueeTextView(context);
            autoMarqueeTextView2 = new AutoMarqueeTextView(context);

            autoMarqueeTextView.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT));
            autoMarqueeTextView2.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT));

            autoMarqueeTextView.setEllipsize(TextUtils.TruncateAt.MARQUEE);
            autoMarqueeTextView2.setEllipsize(TextUtils.TruncateAt.MARQUEE);

            this.addView(autoMarqueeTextView);
            this.addView(autoMarqueeTextView2);
        }
    }

    public void setText(String str) {
        // 设置歌词文本
        if (switchLyric) {
            if (hasMeizu) {
                lyricTextView.setTextT(str);
            } else {
                autoMarqueeTextView.setText(str);
            }
            switchLyric = false;
        } else {
            if (hasMeizu) {
                lyricTextView2.setTextT(str);
            } else {
                autoMarqueeTextView2.setText(str);
            }
            switchLyric = true;
        }

        this.showNext();
    }

    public void setWidth(int i) {
        if (hasMeizu) {
            lyricTextView.setWidth(i);
            lyricTextView2.setWidth(i);
        } else {
            autoMarqueeTextView.setWidth(i);
            autoMarqueeTextView2.setWidth(i);
        }
    }

    public void setTextColor(int i) {
        if (hasMeizu) {
            lyricTextView.setTextColor(i);
            lyricTextView2.setTextColor(i);
        } else {
            autoMarqueeTextView.setTextColor(i);
            autoMarqueeTextView2.setTextColor(i);
        }
    }

    public void setSourceText(CharSequence charSequence) {
        if (hasMeizu) {
            lyricTextView.setText(charSequence);
            lyricTextView2.setText(charSequence);
        } else {
            autoMarqueeTextView.setText(charSequence);
            autoMarqueeTextView2.setText(charSequence);
        }
    }

    public void setSpeed(float f) {
        if (hasMeizu) {
            lyricTextView.setSpeed(f);
            lyricTextView2.setSpeed(f);
        }
    }

    public CharSequence getText() {
        if (switchLyric) {
            if (hasMeizu) {
                return lyricTextView.getText();
            } else {
                return autoMarqueeTextView.getText();
            }
        } else {
            if (hasMeizu) {
                return lyricTextView2.getText();
            } else {
                return autoMarqueeTextView2.getText();
            }
        }
    }

    public TextPaint getPaint() {
        if (switchLyric) {
            if (hasMeizu) {
                return lyricTextView.getPaint();
            } else {
                return autoMarqueeTextView.getPaint();
            }
        } else {
            if (hasMeizu) {
                return lyricTextView2.getPaint();
            } else {
                return autoMarqueeTextView2.getPaint();
            }
        }
    }

    public void setHeight(int i) {
        if (hasMeizu) {
            lyricTextView.setHeight(i);
            lyricTextView2.setHeight(i);
        } else {
            autoMarqueeTextView.setHeight(i);
            autoMarqueeTextView2.setHeight(i);
        }
    }

    public void setTypeface(Typeface typeface) {
        if (hasMeizu) {
            lyricTextView.setTypeface(typeface);
            lyricTextView2.setTypeface(typeface);
        } else {
            autoMarqueeTextView.setTypeface(typeface);
            autoMarqueeTextView2.setTypeface(typeface);
        }
    }

    public void setTextSize(int i, float f) {
        if (hasMeizu) {
            lyricTextView.setTextSize(i, f);
            lyricTextView2.setTextSize(i, f);
        } else {
            autoMarqueeTextView.setTextSize(i, f);
            autoMarqueeTextView2.setTextSize(i, f);
        }
    }

    public void setMargins(int i, int i1, int i2, int i3) {
        if (hasMeizu) {
            FrameLayout.LayoutParams lyricParams = (FrameLayout.LayoutParams) lyricTextView.getLayoutParams();
            lyricParams.setMargins(i, i1, i2, i3);
            lyricTextView.setLayoutParams(lyricParams);

            FrameLayout.LayoutParams lyricParams2 = (FrameLayout.LayoutParams) lyricTextView.getLayoutParams();
            lyricParams2.setMargins(i, i1, i2, i3);
            lyricTextView2.setLayoutParams(lyricParams2);
        } else {
            FrameLayout.LayoutParams lyricParams = (FrameLayout.LayoutParams) autoMarqueeTextView.getLayoutParams();
            lyricParams.setMargins(i, i1, i2, i3);
            autoMarqueeTextView.setLayoutParams(lyricParams);

            FrameLayout.LayoutParams lyricParams2 = (FrameLayout.LayoutParams) autoMarqueeTextView2.getLayoutParams();
            lyricParams2.setMargins(i, i1, i2, i3);
            autoMarqueeTextView2.setLayoutParams(lyricParams2);
        }
    }

    public void setMarqueeRepeatLimit(int i) {
        if (!hasMeizu) {
            autoMarqueeTextView.setMarqueeRepeatLimit(i);
            autoMarqueeTextView2.setMarqueeRepeatLimit(i);
        }
    }

    public void setSingleLine(boolean bool) {
        if (hasMeizu) {
            lyricTextView.setSingleLine(bool);
            lyricTextView2.setSingleLine(bool);
        } else {
            autoMarqueeTextView.setSingleLine(bool);
            autoMarqueeTextView2.setSingleLine(bool);
        }
    }

    public void setMaxLines(int i) {
        if (hasMeizu) {
            lyricTextView.setMaxLines(i);
            lyricTextView2.setMaxLines(i);
        } else {
            autoMarqueeTextView.setMaxLines(i);
            autoMarqueeTextView2.setMaxLines(i);
        }
    }
}

File: app/src/main/java/statusbar/lyric/view/LyricTextSwitchView.kt
package statusbar.lyric.view

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Typeface
import android.text.TextPaint
import android.text.TextUtils
import android.widget.LinearLayout
import android.widget.ViewFlipper
import statusbar.lyric.utils.LogUtils

@SuppressLint("ViewConstructor")
class LyricTextSwitchView(context: Context?, private var hasMeizu: Boolean) : ViewFlipper(context) {
    private var lyricTextView: LyricTextView = LyricTextView(context)
    private var lyricTextView2: LyricTextView = LyricTextView(context)
    private var autoMarqueeTextView: AutoMarqueeTextView = AutoMarqueeTextView(context)
    private var autoMarqueeTextView2: AutoMarqueeTextView = AutoMarqueeTextView(context)
    var switchLyric = false

    init {
        if (hasMeizu) {
            lyricTextView.layoutParams =
                LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
            lyricTextView2.layoutParams =
                LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
            this.addView(lyricTextView)
            this.addView(lyricTextView2)
        } else {
            autoMarqueeTextView.layoutParams =
                LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
            autoMarqueeTextView2.layoutParams =
                LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
            autoMarqueeTextView.ellipsize = TextUtils.TruncateAt.MARQUEE
            autoMarqueeTextView2.ellipsize = TextUtils.TruncateAt.MARQUEE
            this.addView(autoMarqueeTextView)
            this.addView(autoMarqueeTextView2)
        }
    }

    fun setText(str: String) {
        LogUtils.e("设置动画歌词 $str | $switchLyric")
        // 设置歌词文本
        if (switchLyric) {
            if (hasMeizu) {
                lyricTextView.setTextT(str)
            } else {
                autoMarqueeTextView.text = str
            }
            switchLyric = false
        } else {
            if (hasMeizu) {
                lyricTextView2.setTextT(str)
            } else {
                autoMarqueeTextView2.text = str
            }
            switchLyric = true
        }
        showNext()
    }

    fun setWidth(i: Int) {
        if (hasMeizu) {
            lyricTextView.width = i
            lyricTextView2.width = i
        } else {
            autoMarqueeTextView.width = i
            autoMarqueeTextView2.width = i
        }
    }

    fun setTextColor(i: Int) {
        if (hasMeizu) {
            lyricTextView.setTextColor(i)
            lyricTextView2.setTextColor(i)
        } else {
            autoMarqueeTextView.setTextColor(i)
            autoMarqueeTextView2.setTextColor(i)
        }
    }

    fun setSourceText(charSequence: CharSequence?) {
        LogUtils.e("设置歌词 $charSequence")
        if (hasMeizu) {
            lyricTextView.text = charSequence
            lyricTextView2.text = charSequence
        } else {
            autoMarqueeTextView.text = charSequence
            autoMarqueeTextView2.text = charSequence
        }
    }

    fun setSpeed(f: Float) {
        if (hasMeizu) {
            lyricTextView.setSpeed(f)
            lyricTextView2.setSpeed(f)
        }
    }

    val text: CharSequence
        get() = if (switchLyric) {
            if (hasMeizu) {
                lyricTextView.text
            } else {
                autoMarqueeTextView.text
            }
        } else {
            if (hasMeizu) {
                lyricTextView2.text
            } else {
                autoMarqueeTextView2.text
            }
        }
    val paint: TextPaint
        get() = if (switchLyric) {
            if (hasMeizu) {
                lyricTextView.paint
            } else {
                autoMarqueeTextView.paint
            }
        } else {
            if (hasMeizu) {
                lyricTextView2.paint
            } else {
                autoMarqueeTextView2.paint
            }
        }

    fun setHeight(i: Int) {
        if (hasMeizu) {
            lyricTextView.height = i
            lyricTextView2.height = i
        } else {
            autoMarqueeTextView.height = i
            autoMarqueeTextView2.height = i
        }
    }

    fun setTypeface(typeface: Typeface?) {
        if (hasMeizu) {
            lyricTextView.typeface = typeface
            lyricTextView2.typeface = typeface
        } else {
            autoMarqueeTextView.typeface = typeface
            autoMarqueeTextView2.typeface = typeface
        }
    }

    fun setTextSize(i: Int, f: Float) {
        if (hasMeizu) {
            lyricTextView.setTextSize(i, f)
            lyricTextView2.setTextSize(i, f)
        } else {
            autoMarqueeTextView.setTextSize(i, f)
            autoMarqueeTextView2.setTextSize(i, f)
        }
    }

    fun setMargins(i: Int, i1: Int, i2: Int, i3: Int) {
        if (hasMeizu) {
            val lyricParams = lyricTextView.layoutParams as LayoutParams
            lyricParams.setMargins(i, i1, i2, i3)
            lyricTextView.layoutParams = lyricParams
            val lyricParams2 = lyricTextView.layoutParams as LayoutParams
            lyricParams2.setMargins(i, i1, i2, i3)
            lyricTextView2.layoutParams = lyricParams2
        } else {
            val lyricParams = autoMarqueeTextView.layoutParams as LayoutParams
            lyricParams.setMargins(i, i1, i2, i3)
            autoMarqueeTextView.layoutParams = lyricParams
            val lyricParams2 = autoMarqueeTextView2.layoutParams as LayoutParams
            lyricParams2.setMargins(i, i1, i2, i3)
            autoMarqueeTextView2.layoutParams = lyricParams2
        }
    }

    fun setMarqueeRepeatLimit(i: Int) {
        if (!hasMeizu) {
            autoMarqueeTextView.marqueeRepeatLimit = i
            autoMarqueeTextView2.marqueeRepeatLimit = i
        }
    }

    fun setSingleLine(bool: Boolean) {
        if (hasMeizu) {
            lyricTextView.isSingleLine = bool
            lyricTextView2.isSingleLine = bool
        } else {
            autoMarqueeTextView.isSingleLine = bool
            autoMarqueeTextView2.isSingleLine = bool
        }
    }

    fun setMaxLines(i: Int) {
        if (hasMeizu) {
            lyricTextView.maxLines = i
            lyricTextView2.maxLines = i
        } else {
            autoMarqueeTextView.maxLines = i
            autoMarqueeTextView2.maxLines = i
        }
    }
}